### DISCUSSION TITLE
SQL result prints in revealjs format with knitr engine?

### DISCUSSION THREAD

**Discussion by eitsupi** [quarto-dev/quarto-cli#5869] [Q&A]
### Description

Sorry I don't have a complete example yet, but when I run the SQL code block using the knitr engine in revealjs format, the display doesn't seem to be what I intended.
Do you have a sample slide that uses SQL?

![image](https://github.com/quarto-dev/quarto-cli/assets/50911393/f7f32f16-ef94-4330-ac45-c8461f214924)

````qmd
```{r}
#| include: false
#| cashe: false
con_sample <- DBI::dbConnect(duckdb::duckdb())
```

```{sql}
#| connection: con_sample
SELECT *
FROM read_csv_auto('starwars.csv', header=true)
LIMIT 5
```
````

**Comment by cderv** [comment]
What is the intended display ? 

Here is an example 

````markdown
---
format: revealjs
---

```{r}
#| include: false
#| cache: false
con_sample <- dbplyr::nycflights13_sqlite(path = NULL) 
```

# Table

```{sql}
#| connection: con_sample
SELECT *
FROM airlines
LIMIT 5
```
````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/d1ff0ab9-1888-4011-bd83-efcd376da4b7)

**Reply by eitsupi** [reply]
Thanks for your reply.

This problem occurred during the editing of the following slide. Eventually I gave up on the executable code block because the deadline was looming.
https://github.com/eitsupi/tokyorslide/blob/0b02c4bffaac9ccc836ccdaa3f804c7c0d9af28f/tokyor_106/index.qmd#L125-L142

Sorry my screenshot is just a part of the screen, but it is obviously different from yours as there is no header indicating the number of records. (I would rather have no header, but the problem was that the text was unusually large)
I hope to find the time to create a minimal example soon.

**Reply by cderv** [reply]
I wonder if this could be linked to the database used too... We should try create an example that reproduce yes. That would help. 
Possible a knitr bug where incorrect classes or else is applied, leading to wrong style 🤔

**Reply by eitsupi** [reply]
Here is a reproducible example.
In short, I think this is a problem because the output of the SQL code block is always kable, and the font of the table output in kable is not affected by the slide size.
Is it possible to change the output of the SQL code block to the default of df-print?

The reason there was no header was because the print function was overridden.
I apologize for the confusion.

````qmd
---
title: test
subtitle: test
format:
  revealjs:
    width: 1280
    height: 720
engine: knitr
execute:
  cache: false
  echo: true
---

```{r}
#| include: false
con <- DBI::dbConnect(duckdb::duckdb())
duckdb::duckdb_register(con, "mtcars", mtcars)

# Override the default SQL print function to avoid inserting '<div class="knitsql-table">'
# https://github.com/yihui/xaringan/issues/307
knitr::opts_knit$set(sql.print = \(x) paste(knitr::kable(x, format = "markdown"), collapse = "\n"))
```

## H2

```{r}
mtcars |>
  head(5)
```

```{sql}
#| connection: con
SELECT *
FROM mtcars
LIMIT 5
```
````

**Reply by mcanouil** [reply]
You are going to be interested in:
- #3253

**Reply by cderv** [reply]
> Is it possible to change the output of the SQL code block to the default of df-print?

@eitsupi are you expecting the same output that for a data.frame ? No special table formatting ?  

This `sql` chunk output behavior is a **knitr** feature, so maybe some adjustments need to be made at **knitr** level. Currently, the output is either 

* using `kable()` 
* or using `cat(opts_knit$get('sql.print')(data))` 

So not really flexible. To get the default `df_print`, it would need to be mimicked. No easy way to access it for now. The SQL engine is decoupled from other data.frame printing. Definitely not a good choice at the time IMO. 

> `# Override the default SQL print function to avoid inserting '<div class="knitsql-table">'`

If you don't want the div, then you can deactivate setting `knitr.sql.html_div` R option to `FALSE`

The div is there to be able to apply custom styling though. I am aware of one CSS rule used in Quarto which target it. 

> the display doesn't seem to be what I intended.

I think I just need clarification on what you intend to know how to improve. Could be either Quarto side or knitr side. 

Thanks.

**Reply by eitsupi** [reply]
@cderv Sorry for the confusion in my explanation, but anyway the problem here is that the output from the SQL code block has a constant font size and does not look good when the body text font size is changed. So I could not use the executable code block.

I really just wanted just a printout of the results as well as a data frame, no kable needed. But I couldn't figure out how to disable kable or table formatting.

**Reply by mcanouil** [reply]
Would any of the following do?

| default + css | kable + css |
|---------------|-------------|
| <img width="579" alt="Screenshot 2023-06-20 at 11 48 13" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/2f938cea-55e4-4254-bf42-0b923a6b2de0"> | <img width="579" alt="Screenshot 2023-06-20 at 11 48 20" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/ba5b6db5-27d0-4788-af5e-624943e2093c"> |

````qmd
---
title: test
subtitle: test
format:
  revealjs:
    width: 1280
    height: 720
engine: knitr
execute:
  cache: false
  echo: true
include-in-header:
  - text: |
      <style>
      .knitsql-table {
        font-size: 0.5em;
      }
      </style>
---

```{r}
#| include: false
con <- DBI::dbConnect(duckdb::duckdb())
duckdb::duckdb_register(con, "mtcars", mtcars)
```

## Slide 1

```{r}
mtcars |>
  head(5)
```

```{sql}
#| connection: con
SELECT *
FROM mtcars
LIMIT 5
```

## Slide 2

```{r}
#| echo: false
# Override the default SQL print function to avoid inserting '<div class="knitsql-table">'
# https://github.com/yihui/xaringan/issues/307
knitr::opts_knit$set(
  sql.print = function(x) {
    paste(
      knitr::kable(
        x = x,
        format = "html",
        table.attr = 'class="knitsql-table"',
      ),
      collapse = "\n"
    )
  }
)
```

```{sql}
#| connection: con
SELECT *
FROM mtcars
LIMIT 5
```
````

**Reply by eitsupi** [reply]
@mcanouil That looks good, thanks!
What settings are needed to change the font size of a table came from `kable()`? (wondering if that is documented)

**Reply by mcanouil** [reply]
It depends, by default kable outputs markdown, so the font size is the one from the slide.
You can try the following, and you sill see the same display.

````qmd
## Slide 3

|  mpg| cyl| disp|  hp| drat|    wt|  qsec| vs| am| gear| carb|
|----:|---:|----:|---:|----:|-----:|-----:|--:|--:|----:|----:|
| 21.0|   6|  160| 110| 3.90| 2.620| 16.46|  0|  1|    4|    4|
| 21.0|   6|  160| 110| 3.90| 2.875| 17.02|  0|  1|    4|    4|
| 22.8|   4|  108|  93| 3.85| 2.320| 18.61|  1|  1|    4|    1|
| 21.4|   6|  258| 110| 3.08| 3.215| 19.44|  1|  0|    3|    1|
| 18.7|   8|  360| 175| 3.15| 3.440| 17.02|  0|  0|    3|    2|
````

To change it only for the table, you need to make it unique, either by adding a class or by nested it in a div, but this will also affect the source code.
I would have use `class-output`, but `knitr` seems to ignore it for `sql` ...

````
::: {.knitsql-table}

```{sql}
#| connection: con
SELECT *
FROM mtcars
LIMIT 5
```

:::
````

Regarding `kable()`, see <https://bookdown.org/yihui/rmarkdown-cookbook/kable.html>.
There is also `kableExtra` for more customisation, see <https://haozhu233.github.io/kableExtra/>.

**Reply by cderv** [reply]
Sorry it was no clear in my answer above: https://github.com/quarto-dev/quarto-cli/discussions/5869#discussioncomment-6217879. 
or maybe I missed something. I did not understood the issue was just changing CSS rule. 

The default behavior is to add a div of class `knitsql-table` already. This is used to easily style content within the div. 
If the solution is setting some CSS rules, then it could be used to that. 

@eitsupi in your example you are explicitly trying to remove it. and @mcanouil I don't quite understand why you also overwrite `sql.print` to read the class on a kable. 

I see some question about Markdown Table. If the aim is to output HTML for `kable()`, the option can be set. 
```
options(knitr.table.format = "html")
```

 > but anyway the problem here is that the output from the SQL code block has a constant font size and does not look good when the body text font size is changed

I tried to find where the fixed font size was, but I did not found it. Did you identified the rule responsible ? I am probably still missing something, as changing the font size globally is also changing the one from the table. 

I am still a bit lost on what should be improve to help your use case. Sorry.

**Reply by mcanouil** [reply]
> @mcanouil I don't quite understand why you also overwrite sql.print to read the class on a kable.

There are two different examples based on provided code, either by tweaking the CSS without changing the default print method, or by doing both (note that I choose the same name class as the one from `knitr` but it would have been better to use another one).
`knitr` does not add class to markdown table for some reason or to output as it should when using. `class-output`.

Indeed `options(knitr.table.format = "html")` can be used to change the behaviour of all calls to `kable` but this is not necessarily wanted here.


> I am still a bit lost on what should be improve to help your use case. Sorry.

I believe the issue was that kable markdown output was using the default font size from the slide leading to a very big table, while other outputs have their own font size.
Correct me if I am wrong @eitsupi.

**Reply by cderv** [reply]
> knitr does not add class to markdown table for some reason or to output as it should when using. class-output.

sql engine has a specific way of outputing content - it was designed like that and would need a rewrite to change. 

> Indeed options(knitr.table.format = "html") can be used to change the behaviour of all calls to kable but this is not necessarily wanted here

You can apply the option to one chunk only if needed by using the `R.options` chunk option. 

> I believe the issue was that kable markdown output was using the default font size from the slide leading to a very big table, while other outputs have their own font size.

So this would be a more generic issue not realted to **knitr** and `sql` engine then ? This is about table output using the same font size as text on slides ? 

Which now I understand relates to : I don't want tables as output for my `sql` engine. But as I said this is a **knitr** issue. Only way to overwrite it for now is to set  `sql.print` to something else. Another thing that would require a `eng_sql` redesign. 

Easy solution is to use two chunks:

1. Output the SQL result to a variable, 
2. Print the variable

````markdown
---
title: test
subtitle: test
format: revealjs
execute:
  echo: true
---

```{r}
#| include: false
con <- DBI::dbConnect(duckdb::duckdb())
duckdb::duckdb_register(con, "mtcars", mtcars)
```

## Slide 2

```{sql}
#| connection: con
#| output.var: tab
SELECT *
FROM mtcars
LIMIT 5
```

```{r}
#| echo: false
tab
```
````

Otherwise, (weird) example on how  to hack knitr sql output to get a code like output

````markdown
---
title: test
subtitle: test
format: revealjs
execute:
  echo: true
---

```{r}
#| include: false
con <- DBI::dbConnect(duckdb::duckdb())
duckdb::duckdb_register(con, "mtcars", mtcars)
```

## Slide 2

```{r}
#| echo: false
knitr::opts_knit$set(
  sql.print = function(x) {
    paste(c(
      "```",
      capture.output(print(x)),
      "```"),
      collapse = "\n"
    )
  }
)
```

```{sql}
#| connection: con
SELECT *
FROM mtcars
LIMIT 5
```

```{r}
#| echo: false
DBI::dbDisconnect(con)
```

````

Hope this times, it helps

**Reply by mcanouil** [reply]
The thread mentioned in https://github.com/quarto-dev/quarto-cli/discussions/5869#discussioncomment-6196425 also make use of the `output.var` trick (and it was marked as the answer). (It's weird i never used that ...)

- https://github.com/quarto-dev/quarto-cli/discussions/3253#discussioncomment-4127342
